### PR TITLE
[MI-2081] Add feature to delete and disconnect user on encryption sec…

### PR DESCRIPTION
**Pull request by ayusht2810**
…ret change

### CODE REVIEW

File: server/store/utils.go [hunk 1]
```diff
@@ -145,3 +146,11 @@ func GetKeyHash(key string) string {
 	_, _ = hash.Write([]byte(key))
 	return base64.StdEncoding.EncodeToString(hash.Sum(nil))
 }
+
+func IsValidUserKey(key string) (string, bool) {
+	res := strings.Split(key, "_")
+	if res[0] == constants.UserIDPrefix {
```
Comment (Kshitij-Katiyar) [inline server/store/utils.go:152]: not checking the length of res here

File: server/store/store.go [hunk 2]
```diff
@@ -58,3 +59,35 @@ func (s *Store) Delete(key string) error {
...
+	for {
+		kvList, err := s.api.KVList(page, constants.UsersPerPage)
+		if err != nil {
+			return err
+		}
+		if len(kvList) == 0 {
+			return nil
+		}
+
+		isUserDeleted := false
```
Comment (Kshitij-Katiyar) [inline server/store/store.go:74]: please explain the need for this flag and why we are updating the pagnumber according to this flag

File: server/plugin/hooks.go [hunk 3]
```diff
@@ -26,13 +26,20 @@ func (p *Plugin) OnConfigurationChange() error {
...
+	oldEncryptionSecret := p.getConfiguration().EncryptionSecret
 	mattermostSiteURL := p.API.GetConfig().ServiceSettings.SiteURL
 	if mattermostSiteURL == nil {
 		return errors.New("plugin requires Mattermost Site URL to be set")
 	}
 	configuration.MattermostSiteURL = *mattermostSiteURL
+	p.setConfiguration(configuration)
+
+	if oldEncryptionSecret != "" && oldEncryptionSecret != p.getConfiguration().EncryptionSecret {
+		if err := p.Store.DeleteUserOnSecretChange(); err != nil {
```
Comment (Kshitij-Katiyar) [inline server/plugin/hooks.go:38]: we are deling whole user or just its token ?

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/hooks.go (hunk 3 above)
Comment (ayusht2810) [reply server/plugin/hooks.go:38]: we are deleting details related to OAuth, like access token, refresh token

Comment (Kshitij-Katiyar) [reply server/plugin/hooks.go:38]: then DeleteUserTokenOnSecretChange is a better name for that function then

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/hooks.go [hunk 4]
```diff
@@ -26,13 +26,20 @@ func (p *Plugin) OnConfigurationChange() error {
...
+	oldEncryptionSecret := p.getConfiguration().EncryptionSecret
 	mattermostSiteURL := p.API.GetConfig().ServiceSettings.SiteURL
 	if mattermostSiteURL == nil {
 		return errors.New("plugin requires Mattermost Site URL to be set")
 	}
 	configuration.MattermostSiteURL = *mattermostSiteURL
+	p.setConfiguration(configuration)
+
+	if oldEncryptionSecret != "" && oldEncryptionSecret != p.getConfiguration().EncryptionSecret {
+		if err := p.Store.DeleteUserTokenOnSecretChange(); err != nil {
```
Comment (avas27JTG) [inline server/plugin/hooks.go:38]:
```suggestion
		if err := p.Store.DeleteUserTokenOnEncryptionSecretChange(); err != nil {
```

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**